test/unit/event.js

```diff
@@ -3041,6 +3041,49 @@ QUnit.test( "focus-blur order (#12868)", function( assert ) {
 	}, 50 );
 } );
 
+QUnit.test( "Event handling works with multiple async focus events (gh-4350)", function( assert ) {
+	assert.expect( 3 );
+
+	var remaining = 3,
+		input = jQuery( "#name" ),
+
+		// Support: IE <=9 - 11+
+		// focus and blur events are asynchronous; this is the resulting mess.
+		// The browser window must be topmost for this to work properly!!
+		done = assert.async();
+
+	input
+		.on( "focus", function() {
+			remaining--;
+			assert.ok( true, "received focus event, expecting " + remaining + " more" );
+			if ( remaining > 0 ) {
+				input.trigger( "blur" );
+			} else {
+				done();
+			}
+		} )
+		.on( "blur", function() {
+			setTimeout( function() {
+				input.trigger( "focus" );
+			} );
+		} );
+
+	// gain focus
+	input.trigger( "focus" );
+
+	// DOM focus is unreliable in TestSwarm
+	setTimeout( function() {
```

Comment:
Shouldn't the `setTimeout` be given some positive timeout value so that this fallback is not executed too fast?

Also, do you think there might be a way to avoid this fallback?

Reply:
I don't see how to avoid this fallback, but a specific timeout is not necessary because queueing itself ensures that the first focus handler will run before this callback if it is to be invoked at all.

Reply:
Makes sense.

BTW, if this is unreliable only in TestSwarm the you could guard the fallback with an additional `QUnit.isSwarm`or whatever it’s named, then it’d be guaranteed it’s not skipped in our Karma tests in Chrome headless.

Reply:
Good idea; done.